app/src/main/java/com/infomaniak/drive/ui/menu/GalleryViewModel.kt

```diff
@@ -17,69 +17,87 @@
+import com.infomaniak.drive.data.api.CursorApiResponse
 import com.infomaniak.drive.data.cache.FileController
 import com.infomaniak.drive.data.models.File
 import com.infomaniak.drive.utils.IsComplete
-import com.infomaniak.drive.utils.isLastPage
-import com.infomaniak.lib.core.models.ApiResponse
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
 
 class GalleryViewModel : ViewModel() {
-    private var getGalleryJob: Job = Job()
 
-    var lastGalleryPage = 1
-    var lastGalleryLastPage = 1
+    private var getGalleryJob: Job? = null
 
-    val loadMoreGallery = MutableLiveData<Pair<Int, Boolean>>()
-    val galleryApiResult = loadMoreGallery.switchMap { (driveId, ignoreCloud) ->
-        getLastGallery(driveId, ignoreCloud)
+    private var currentCursor: String? = null
+
+    val galleryApiResult = MutableLiveData<Pair<ArrayList<File>, IsComplete>?>()
+
+    fun loadLastGallery(driveId: Int, ignoreCloud: Boolean) {
+        getGalleryJob?.cancel()
+        getGalleryJob = viewModelScope.launch(Dispatchers.IO) {
+            galleryApiResult.postValue(getLastGallery(driveId, ignoreCloud, isFirstPage = true))
+        }
+    }
+
+    fun loadMoreGallery(driveId: Int, ignoreCloud: Boolean) {
+        currentCursor?.let {
+            getGalleryJob?.cancel()
+            getGalleryJob = viewModelScope.launch(Dispatchers.IO) {
+                galleryApiResult.postValue(getLastGallery(driveId, ignoreCloud, isFirstPage = false, currentCursor))
+            }
+        }
     }
 
     private fun getLastGallery(
         driveId: Int,
         ignoreCloud: Boolean = false,
-    ): LiveData<Pair<ArrayList<File>, IsComplete>?> {
-        getGalleryJob.cancel()
+        isFirstPage: Boolean = true,
+        cursor: String? = null,
+    ): Pair<ArrayList<File>, IsComplete>? {
+        getGalleryJob?.cancel()
         getGalleryJob = Job()
 
-        return liveData(Dispatchers.IO + getGalleryJob) {
-            if (ignoreCloud) emitRealmGallery() else fetchApiGallery(driveId)
-        }
+        return if (ignoreCloud) emitRealmGallery() else fetchApiGallery(driveId, isFirstPage, cursor)
     }
 
-    private suspend fun LiveDataScope<Pair<ArrayList<File>, IsComplete>?>.emitRealmGallery() {
-        emit(FileController.getGalleryDrive() to true)
+    private fun emitRealmGallery(): Pair<ArrayList<File>, Boolean> {
+        currentCursor = null
+        return FileController.getGalleryDrive() to true
     }
 
-    private suspend fun LiveDataScope<Pair<ArrayList<File>, IsComplete>?>.fetchApiGallery(driveId: Int) {
-        val page = lastGalleryPage
-        val apiResponse = ApiRepository.getLastGallery(driveId = driveId, page = page)
-        if (apiResponse.isSuccess()) emitApiGallery(apiResponse, page) else emitRealmGallery()
+    private fun fetchApiGallery(driveId: Int, isFirstPage: Boolean, cursor: String?): Pair<java.util.ArrayList<File>, Boolean>? {
```

Comment:
Any reasons to use the java ArrayList instead of the kotlin one, or is it just a oversight ?